### PR TITLE
[SPARK-31990][SS] Use toSet.toSeq in Dataset.dropDuplicates

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This PR partially revert SPARK-31292 in order to provide a hot-fix for a bug in `Dataset.dropDuplicates`; we must preserve the input order of `colNames` for `groupCols` because the Streaming's state store depends on the `groupCols` order.

### Why are the changes needed?

Bug fix.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added tests in `DataFrameSuite`.

### CODE REVIEW

**Comment by maropu** [comment]
cc: @xuanyuanking @HeartSaVioR @srowen @gatorsmile

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2554]: @xuanyuanking Do you have a test input seq to reproduce this issue? I checked if this test suite could fail w/o the hot-fix, but it passed...

**Comment by dongjoon-hyun** [comment]
Thank you for making a swift fix, @maropu .
cc @dbtsai and @holdenk

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2554]: ?

**Comment by xuanyuanking** [comment]
Yes, this incompatible bug is found by a WIP validation logic. I will reply
the details and reference the PR soon.

Dongjoon Hyun <notifications@github.com>于2020年6月15日 周一09:54写道：

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2554]: I just couldn't find a test query to reproduce the issue described in the JIRA. I think @xuanyuanking can provide it because he find this issue...

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2548]: This looks logically the same with Scala's `distinct` implementation. Is there a change?

**Comment by HeartSaVioR** [comment]
Let's use `[SS]` instead as it's specific to SS issue.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2554]: You may want to find the case where `toSet.toSeq` produces different order than `distinct`.

And then you may need to craft a streaming query doing dropDuplicate with having grouping key just you found, and run the query with Spark 2.4.x (where SPARK-31292 is not applied), and store checkpoint, and run the query with Spark 3.x (where SPARK-31292 is applied) with restoring checkpoint.

There're a couple of test suites doing that - one of example is https://github.com/apache/spark/blob/8282bbf12d4e174986a649023ce3984aae7d7755/sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala#L880-L952

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2548]: Ah, I see and it looks the same: https://github.com/scala/scala/blob/2.12.x/src/library/scala/collection/SeqLike.scala#L504-L523 But, (I'm not a Scala compiler expert though),  the `distinct` scala API makes sure an input order is preserved? I read [the @HeartSaVioR comment ](https://issues.apache.org/jira/browse/SPARK-31990?focusedCommentId=17135286&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17135286) in the Jira, then I'm not sure about that. So, I wrote this method to preserve the order explicitly for safeguard.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2554]: Technically, as you're implementing the same as `distinct` does, it doesn't fix the issue for restoring from old checkpoint. Both Spark 3.x and your patch will fail. That's what I commented in JIRA - "do the right fix" vs "leave it as it is".

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2548]: I think Scala wouldn't change the implementation unless there's strong reason to do so, but in Scala 2.13 they removed the orderness guarantee on the method doc in `distinct`. Till Scala 2.12 it clearly explains the return value as preserving the "first occurrence", and the explanation no longer exists in Scala 2.13.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2548]: So consider this as manual implementation of distinct so that we don't even get affected by Scala changes.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2548]: Oh I didn't see @maropu 's comment while I'm commenting. ;) Thanks for explaining.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2548]: The reported issue claims that Scala `distinct` function was not enough. That's the reason why I asked that `What is the difference to fix Spark issue`.

As @HeartSaVioR 's commented (https://github.com/apache/spark/pull/28830#discussion_r439904302), we need a different code.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2548]: Ah, I see. Yea, I'll update the code based on `toSeq.toSeq`.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2554]: BTW, @HeartSaVioR . Is there a test case failure using the same Spark version checkpointing? I'm curious if this only occurs between different Spark versions.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2548]: How about simply revert this line to https://github.com/apache/spark/pull/28062/files#diff-7a46f10c3cedbf013cf255564d9483cdL2458, use the original implementation of `toSet`.
Yes, the `toSet.toSeq` might incompatible during to Scala version, but I think the current fix should just keep the original order. How to detect the order changing and have solid validation should be the work of SPARK-31894 and SPARK-27237. WDYT?

**Comment by dongjoon-hyun** [comment]
Hi, All.
This issue is marked as a hotfix for the blocker issue, but the validation of this issue looks non-trivial. Since `toSet.toSeq` is used since Apache Spark 2.2.0 (SPARK-19497) and SPARK-31292 is just an `Improvement` with `Trivial` issue. I'd like to propose to revert SPARK-31292 from `branch-3.0` first? We will keep SPARK-31292 in `master` branch still and proceed this PR to find a better way for Apache Spark 3.1.0.

I know that the reverting is not a good solution for the original author as mentioned by @HeartSaVioR in the dev mailing list, but I believe that is the proper way in this case to cut Apache Spark 3.0.1. How do you think about that?

**Comment by gatorsmile** [comment]
Yes. I prefer to reverting the original fix in 3.0.1. and then discuss how to solve/avoid the problems in a proper way.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala
Comment (xuanyuanking) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2554]: Thanks for adding a new UT here. Since this issue was found when investigating the test failure in https://github.com/apache/spark/pull/28707#issuecomment-639861273,
how about reusing the UT `default config of includeHeader doesn't break existing query from Spark 2.4` in `KafkaMicroBatchV2SourceSuite`? I think we don't need to add a new UT for this regression after #28707. That is to say after #28707 is merged, if we don't do the fix, the mentioned UT will fail.

**Comment by maropu** [comment]
okay, I'll revert that part in this PR first.

**Review by xuanyuanking — COMMENTED** [review]
Thanks for the quick fix @maropu! I think maybe we can simplify the bugfix by combining it together with #28707. WDYT? I'll also reference this PR with #28707.

**Comment by HeartSaVioR** [comment]
+1 to partial revert which should be also OK with author. (I guess it was applied simply by pattern, and it wasn’t for some outstanding improvement, so no problem for author as well.)

**Comment by xuanyuanking** [comment]
Yep, I think just revert that part is good enough. I will give more context and details on #28707.

**Comment by dongjoon-hyun** [comment]
Ya. +1 for partial revert in this PR.

**Comment by maropu** [comment]
> Thanks for the quick fix @maropu! I think maybe we can simplify the bugfix by combining it together with #28707. WDYT? I'll also reference this PR with #28707.

@xuanyuanking yea, looks fine to me. Could you takes this over? Thanks, anyway!

**Review by xuanyuanking — APPROVED** [review]
@maropu Sure, I'm writing comments in #28707. Will cc all of us and reference it with SPARK-31990 and this PR soon.

**Comment by HeartSaVioR** [comment]
How we plan to consolidate both? How we will write JIRA title/description and PR title/description? Which is the type of the consolidated issue? Is the consolidated issue a blocker?

Things would be simple if we merge the partial fix as it is, and spend our efforts to discuss how to guide known issue - this is one of candidates for Spark 3.0.0. This is clearly a bugfix which is a "blocker" preventing some of end users migrate to Spark 3.0.0.

It'd be no harm for #28707 to wait for this patch to be merged, and rebase to fix the test failure.

**Comment by xuanyuanking** [comment]
```
How we plan to consolidate both? How we will write JIRA title/description and PR title/description? Which is the type of the consolidated issue? Is the consolidated issue a blocker?
```
Here's my plan to consolidate both: https://github.com/apache/spark/pull/28707#issuecomment-643916110, this will also comment in JIRA & PR description.
Yes, #28707 is blocking by this fix.

**Comment by HyukjinKwon** [comment]
I am okay to revert it for now but I couldn't fully follow why we expect an explicit order from a set. Has it been ever guaranteed somewhere? Using `distinct`, we can expect the deterministic order but we're reverting back to using a set because of the deterministic order (?).

**Comment by dongjoon-hyun** [comment]
The last commit is to trying to preserve the previous behavior (whatever it was) since Apache Spark 2.2.0 although there is no guarantee which it safe or not. We will revisit the correct issue later after 3.0.1.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
I am okay with that; I was just wondering even the previous behaviour was deterministic or not, and the current change looked righter to me. Given that we're going to revisit anyway, LGTM from me too.

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2546]: I think this is good for now.

In the future, this may still be broken by Scala version upgrade, and hopefully @xuanyuanking 's unsafe row validation can detect it. Then we can change it and use a deterministic order.

Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2546]: Yep, I also mentioned this at https://github.com/apache/spark/pull/28830#discussion_r439909489, we can relay on the validation checking and integrated tests.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2546]: Worth noting that we need to have "concrete" solution eventually - if columns are all having same type neither #28830 nor #24173 catch the change. I roughly remember the similar issue on pyspark, which was trying to fix the issue on order vs name, don't remember how it ended up. cc. @HyukjinKwon

**Review by HeartSaVioR — APPROVED** [review]
LGTM as it is.

**Comment by HeartSaVioR** [comment]
Btw now we know it is broken in Spark 3.0.0, and we will fix it again in Spark 3.0.1. Do we have some best practice to follow on guiding such change to end users?

**Comment by gengliangwang** [comment]
retest this please

**Review by gengliangwang — APPROVED** [review]

**Comment by cloud-fan** [comment]
> Btw now we know it is broken in Spark 3.0.0, and we will fix it again in Spark 3.0.1.

I think we should list it as a known issue of 3.0.0, and release 3.0.1 soon.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2546]: Ah, that was fixed in a way by adding an env variable. That case also was specific to Python 2 which is deprecated now so it's rather a corner case.

**Comment by dongjoon-hyun** [comment]
Thank you all. Merged to master/3.0.

**Comment by maropu** [comment]
Thanks, all!

**Pull request closed without merging**
